### PR TITLE
File name param for direct access url requests

### PR DESCRIPTION
**Pull request by canpan14**
Adds optional file name parameter for all direct access url requests https://github.com/Alfresco/alfresco-community-repo/issues/2064

### CODE REVIEW

**Review by mpichura — CHANGES_REQUESTED** [review]
Nice work but for backwards compatibility please don't replace public interface methods with new ones (see my comments on the interfaces).

**Review by montgolfiere — COMMENTED** [review]
We should also have a corresponding PR ready that documents the new optional param in the OpenAPI Spec, ie. for all related endpoints / schema here:

https://github.com/Alfresco/rest-api-explorer/blob/master/src/main/webapp/definitions/alfresco-core.yaml

File: remote-api/src/main/java/org/alfresco/rest/api/DirectAccessUrlHelper.java [hunk 1]
```diff
@@ -63,4 +63,15 @@ public boolean getAttachment(DirectAccessUrlRequest directAccessUrlRequest)
         }
         return attachment;
     }
+
+    
+    public String getFileName(DirectAccessUrlRequest directAccessUrlRequest)
```
[the remark this answers is not in the document]
Comment (canpan14) [reply remote-api/src/main/java/org/alfresco/rest/api/DirectAccessUrlHelper.java:68]: Sure! I was just following the format of the method above.

**Comment by canpan14** [comment]
> We should also have a corresponding PR ready that documents the new optional param in the OpenAPI Spec, ie. for all related endpoints / schema here:
> 
> https://github.com/Alfresco/rest-api-explorer/blob/master/src/main/webapp/definitions/alfresco-core.yaml

Good callout, I'll get that written up.

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 2]
```diff
@@ -241,11 +241,12 @@ default public DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean
      * @param propertyQName the name of the property, which must be of type <b>content</b>
      * @param attachment {@code true} if an attachment URL is requested, {@code false} for an embedded {@code URL}.
      * @param validFor The time at which the direct access {@code URL} will expire.
+     * @param fileName Optional name for the file when downloaded
      * @return A direct access {@code URL} object for the content.
      * @throws UnsupportedOperationException if the store is unable to provide the information.
      */
     @Auditable(parameters = {"nodeRef", "propertyQName", "validFor"})
-    DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, QName propertyQName, boolean attachment, Long validFor);
+    DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, QName propertyQName, boolean attachment, Long validFor, String fileName);
```
[the remark this answers is not in the document]
Comment (canpan14) [reply repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java:266]: Fixed

File: remote-api/src/main/java/org/alfresco/rest/api/Renditions.java [hunk 3]
```diff
@@ -257,8 +257,9 @@ default DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, String versionI
      * @param renditionId the rendition id
      * @param attachment  {@code true} if an attachment {@code URL} is requested, {@code false} for an embedded {@code URL}
      * @param validFor    the time at which the direct access {@code URL} will expire
+     * @param fileName    optional name for the file when downloaded
      * @return            a direct access {@code URL} object for the content.
      */
-    DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, String versionId, String renditionId, boolean attachment, Long validFor);
+    DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, String versionId, String renditionId, boolean attachment, Long validFor, String fileName);
```
[the remark this answers is not in the document]
Comment (canpan14) [reply remote-api/src/main/java/org/alfresco/rest/api/Renditions.java:276]: Fixed

File: remote-api/src/main/java/org/alfresco/rest/api/DeletedNodes.java [hunk 4]
```diff
@@ -121,8 +121,9 @@ default DirectAccessUrl requestContentDirectUrl(String archivedId, String rendit
      * @param renditionId The rendition id for which to obtain the direct access {@code URL}
      * @param attachment {@code true} if an attachment {@code URL} is requested, {@code false} for an embedded {@code URL}, {@code true} by default.
      * @param validFor The time at which the direct access {@code URL} will expire.
+     * @param fileName Optional overide for file name
      * @return A direct access {@code URL} object for the content.
      */
-    DirectAccessUrl requestContentDirectUrl(String archivedId, String renditionId, boolean attachment, Long validFor);
+    DirectAccessUrl requestContentDirectUrl(String archivedId, String renditionId, boolean attachment, Long validFor, String fileName);
```
[the remark this answers is not in the document]
Comment (canpan14) [reply remote-api/src/main/java/org/alfresco/rest/api/DeletedNodes.java:139]: Fixed

File: remote-api/src/main/java/org/alfresco/rest/api/Nodes.java [hunk 5]
```diff
@@ -313,17 +313,18 @@ default DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachm
...
     /**
      * Gets a presigned URL to directly access content.
      * @param nodeRef The node reference for which to obtain the direct access {@code URL}
      * @param attachment {@code true} if an attachment {@code URL} is requested, {@code false} for an embedded {@code URL}.
      * @param validFor The time at which the direct access {@code URL} will expire.
+     * @param fileName Optional name for the file when downloaded
      * @return A direct access {@code URL} object for the content.
      */
-    DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachment, Long validFor);
+    DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachment, Long validFor, String fileName);
```
[the remark this answers is not in the document]
Comment (canpan14) [reply remote-api/src/main/java/org/alfresco/rest/api/Nodes.java:339]: Fixed

**Comment by canpan14** [comment]
@montgolfiere PR on that project https://github.com/Alfresco/rest-api-explorer/pull/197

**Comment by canpan14** [comment]
> A couple of test failures now:
> 
> ```
> Error:  testRequestContentDirectUrl_SystemWideIsEnabledButStoreIsDisabled(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.005 s  <<< ERROR!
> java.lang.StackOverflowError
> 
> Error:  testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.005 s  <<< ERROR!
> java.lang.StackOverflowError
> 
> [INFO] 
> [INFO] Results:
> [INFO] 
> Error:  Failures: 
> Error:    ContentServiceImplUnitTest.testRequestContentDirectUrl_SystemWideIsDisabled:138 unexpected exception type thrown; expected:<org.alfresco.repo.content.directurl.DirectAccessUrlDisabledException> but was:<java.lang.StackOverflowError>
> Error:  Errors: 
> Error:    ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreIsEnabledButNotImplemented » StackOverflow
> Error:    ContentServiceImplUnitTest.testRequestContentDirectUrl_SystemWideIsEnabledButStoreIsDisabled » StackOverflow
> ```
> 
> I assume there's an accidental loop with the methods calling themselves now.

Whoops! Did a bad job repointing the previous interfaces. Should be fixed now.

**Review by mpichura — APPROVED** [review]
LGTM

**Pull request merged**